src/arithmetic/comprehension_funcs/comprehension_funcs.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2018-2020 Redis Labs Ltd. and Contributors
+ *
+ * This file is available under the Redis Labs Source Available License Agreement
+ */
+
+#include "comprehension_funcs.h"
+#include "../../RG.h"
+#include "../func_desc.h"
+#include "../arithmetic_expression.h"
+#include "../../value.h"
+#include "../../util/arr.h"
+#include "../../datatypes/array.h"
+#include "../../execution_plan/record.h"
+#include "../../filter_tree/filter_tree.h"
+
+/* Routine for freeing a list comprehension's subtree of arithmetic expressions.
+ * The predicate and eval routines require special handling to be freed properly. */
+void ListComprehension_Free(AR_ExpNode *exp) {
+	// The child at index 0 contains the local variable as a variadic node.
+	AR_ExpNode *variable = exp->op.children[0];
+	ASSERT(variable->type == AR_EXP_OPERAND && variable->operand.type == AR_EXP_CONSTANT);
+	SIValue variable_val = variable->operand.constant;
+	AR_EXP_Free(variable_val.ptrval);
+
+	// The child at index 2 contains the filter tree, if any.
+	AR_ExpNode *predicate = exp->op.children[2];
+	ASSERT(predicate->type == AR_EXP_OPERAND && predicate->operand.type == AR_EXP_CONSTANT);
+	SIValue predicate_val = predicate->operand.constant;
+	// If this list comprehension has a filter tree, free it.
+	if(SI_TYPE(predicate_val) & T_PTR) FilterTree_Free(predicate_val.ptrval);
+
+	// The child at index 3 contains the eval routine, if any.
+	AR_ExpNode *eval = exp->op.children[3];
+	assert(eval->type == AR_EXP_OPERAND && eval->operand.type == AR_EXP_CONSTANT);
+	SIValue eval_val = eval->operand.constant;
+	// If this list comprehension has an eval routine, free it.
+	if(SI_TYPE(eval_val) & T_PTR) AR_EXP_Free(eval_val.ptrval);
+}
+
+/* Routine for cloning a list comprehension's subtree of arithmetic expressions.
+ * The predicate and eval routines require special handling to be cloned properly. */
+void ListComprehension_Clone(AR_ExpNode *orig, AR_ExpNode *clone) {
```

Comment:
Similar comment as above, if an arithmetic function would have a context this can be simplified and the abstraction would be damaged.